deflect/qt/OffscreenQuickView.cpp

```diff
@@ -0,0 +1,247 @@
+/*    OF  THE  USE OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE    */
+/*    POSSIBILITY OF SUCH DAMAGE.                                    */
+/*                                                                   */
+/* The views and conclusions contained in the software and           */
+/* documentation are those of the authors and should not be          */
+/* interpreted as representing official policies, either expressed   */
+/* or implied, of Ecole polytechnique federale de Lausanne.          */
+/*********************************************************************/
+
+#include "OffscreenQuickView.h"
+
+#include "QuickRenderer.h"
+
+#include <QOpenGLContext>
+#include <QOpenGLFunctions>
+#include <QQmlComponent>
+#include <QQmlEngine>
+#include <QQuickItem>
+#include <QQuickRenderControl>
+#include <QThread>
+
+namespace deflect
+{
+namespace qt
+{
+
+OffscreenQuickView::OffscreenQuickView( std::unique_ptr<QQuickRenderControl>
+                                        renderControl, const RenderMode mode )
+    : QQuickWindow{ renderControl.get() }
+    , _renderControl{ std::move( renderControl ) }
+    , _mode{ mode }
+    , _qmlEngine{ new QQmlEngine }
+    , _qmlComponent{ new QQmlComponent{ _qmlEngine.get() }}
+{
+    if( !_qmlEngine->incubationController( ))
+        _qmlEngine->setIncubationController( incubationController( ));
+
+    // remote URL to QML components are loaded asynchronously
+    connect( _qmlComponent.get(), &QQmlComponent::statusChanged,
+             this, &OffscreenQuickView::_setupRootItem );
+
+    // Now hook up the signals. For simplicy we don't differentiate between
+    // renderRequested (only render is needed, no sync) and sceneChanged (polish
+    // and sync is needed too).
+    connect( _renderControl.get(), &QQuickRenderControl::renderRequested,
+             this, &OffscreenQuickView::_requestRender );
+    connect( _renderControl.get(), &QQuickRenderControl::sceneChanged,
+             this, &OffscreenQuickView::_requestRender );
+}
+
+OffscreenQuickView::~OffscreenQuickView()
+{
+    killTimer( _renderTimer );
+    _renderTimer = 0;
+    killTimer( _stopRenderingDelayTimer );
+    _stopRenderingDelayTimer = 0;
+
+    _quickRenderer->stop();
+    if( _quickRendererThread )
+    {
+        _quickRendererThread->quit();
+        _quickRendererThread->wait();
+    }
+
+    // delete first to free scenegraph resources for following destructions
+    _renderControl.reset();
+}
+
+std::future<bool> OffscreenQuickView::load( const QUrl& url )
+{
+    _qmlComponent->loadUrl( url );
+    return _loadPromise.get_future();
+}
+
+QQuickItem* OffscreenQuickView::getRootItem() const
+{
+    return _rootItem.get();
+}
+
+QQmlEngine* OffscreenQuickView::getEngine() const
+{
+    return _qmlEngine.get();
+}
+
+QQmlContext* OffscreenQuickView::getRootContext() const
+{
+    return _qmlEngine->rootContext();
+}
+
+QImage OffscreenQuickView::getImage() const
+{
+    return _quickRenderer->fbo()->toImage();
```

Comment:
You're right, and I realize it is also not very safe that this can only be called from a Qt::DirectConnection (while the FBO is bound). A better way would be to get the image internally just afterRender() and always keep a copy, so that class users don't risk making any mistake.